### PR TITLE
[Enchance] Support random seed for distributed sampler

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation
Modify distributed sampler and keep the manual_seed behavior same with pytorch original version

Refers to https://github.com/open-mmlab/mmdetection/pull/7440/files, https://github.com/open-mmlab/mmdetection/pull/7432/files and https://github.com/open-mmlab/mmediting/pull/815.

## Modification
add seed to distributed sampler and apply `g.manual_seed(self.epoch + self.seed)`
synchronize the seed in init function

### CODE REVIEW

**Review by MeowZheng — COMMENTED** [review]
move mmseg/utils/dist_util.py to mmseg/core/utils/dist_util.py

File: mmseg/datasets/samplers/distributed_sampler.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from __future__ import division
+
+import numpy as np
+import torch
+from torch.utils.data import DistributedSampler as _DistributedSampler
+
+from mmseg.utils import sync_random_seed
```
Comment (MeowZheng) [inline mmseg/datasets/samplers/distributed_sampler.py:8]:
```suggestion
from mmseg.core import sync_random_seed
```

File: mmseg/core/utils/dist_util.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    data in the dataset. Therefore, this function is used to make sure that
+    each rank shuffles the data indices in the same order based
+    on the same seed. Then different ranks could use different indices
+    to select non-overlapped data from the same data list.
+    Args:
+        seed (int, Optional): The seed. Default to None.
+        device (str): The device where the seed will be put on.
+            Default to 'cuda'.
+    Returns:
+        int: Seed to be used.
```
Comment (MeowZheng) [inline mmseg/core/utils/dist_util.py:29]: format these lines

File: mmseg/datasets/samplers/distributed_sampler.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+from mmseg.core.utils import sync_random_seed
+
+
+class DistributedSampler(_DistributedSampler):
+    """DistributedSampler inheriting from
+    `torch.utils.data.DistributedSampler`.
+
+    In pytorch of lower versions, there is no `shuffle` argument. This child
+    class will port one to DistributedSampler.
```
Comment (MeowZheng) [inline mmseg/datasets/samplers/distributed_sampler.py:20]: might remove these lines, `shuffle` has been supported from pytorch 1.2, and mmseg requires pytorch >= 1.5

File: mmseg/datasets/samplers/distributed_sampler.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        assert len(indices) == self.total_size
+
+        # subsample
+        indices = indices[self.rank:self.total_size:self.num_replicas]
+        assert len(indices) == self.num_samples
+
+        return iter(indices)
+
+
+class MixedBatchDistributedSampler(Sampler):
```
Comment (MeowZheng) [inline mmseg/datasets/samplers/distributed_sampler.py:80]: `MixedBatchDistributedSampler` is just for optical flow models training, and there is no need to implement it in mmseg

**Review by MeowZheng — APPROVED** [review]
